internal/fs/dir_handle.go

```diff
@@ -242,21 +266,36 @@ func (dh *dirHandle) ensureEntries(ctx context.Context) (err error) {
 //
 // LOCKS_REQUIRED(dh.Mu)
 // LOCKS_EXCLUDED(du.in)
-func (dh *dirHandle) ReadDir(
+func (dh *DirHandle) ReadDir(
 	ctx context.Context,
 	op *fuseops.ReadDirOp) (err error) {
 	// If the request is for offset zero, we assume that either this is the first
 	// call or rewinddir has been called. Reset state.
+
+	dh.Mu.Lock()
 	if op.Offset == 0 {
 		dh.entries = nil
 		dh.entriesValid = false
+		dh.err = nil
+		dh.fetchOver = false
 	}
+	ev := dh.entriesValid
+	dh.Mu.Unlock()
 
-	//when the offset is zero(new call made means fetch)
-	//when the offset is non zero and len(dh.entries ) - offset <0 means not enough data to serve the call
-	if !dh.entriesValid || (len(dh.entries) <= int(op.Offset) && op.Offset != 0 && ContinuationToken != "") {
-		err = dh.ensureEntries(ctx)
-		if err != nil {
+	if !ev {
+		go dh.FetchEntriesAsync(fuseops.RootInodeID, true)
```

Comment:
how about you put this line after 282, once the go routine is invoked is the lock released?

Reply:
When the go routine is invoked, the lock still remains with the parent.So the child will be blocked until parent releases it.But Since we release the lock shortly after in the parent, it might work.